### PR TITLE
Implement TemplatePartAttribute from WPF

### LINKED ISSUE #7432: Add TemplatePartAttribute for Controls
**Is your feature request related to a problem? Please describe.**

Named parts in control templates are not always clear to control authors, designers and other IDE tooling.

**Describe the solution you'd like**

WPF has the TemplatePart attribute used with templated controls. This indicates the named parts of the template, and their type, that are used within code-behind. This information was historically used in designers (Blend) as well as with automated checks to ensure default styles and their templates are complete.

This TemplatePart attribute should be added to avalonia and used for all controls (example from WPF). This should be used for all elements that start with the "PART_" prefix in the name.

```csharp
[TemplatePart(Name="PART_ContentHost", Type=typeof(FrameworkElement))]
public abstract class TextBoxBase
```

 * [Microsoft Docs TemplatePartAttribute ](https://docs.microsoft.com/en-us/dotnet/api/system.windows.templatepartattribute?redirectedfrom=MSDN&view=windowsdesktop-6.0)
 * [Template parts and custom controls (quick tip)](https://www.jeff.wilcox.name/2010/04/template-part-tips/)
 * [WPF TemplateParts](https://paulstovell.com/wpf-part-names/)

**Describe alternatives you've considered**

The alternative is no attribute at all as currently exists.

**Additional context**

Information for PseuodClasses is provided in another attribute. This one should be included as well for completeness.

### PR DESCRIPTION
**Pull request by robloo**
## What does the pull request do?

Closes #7432 

## What is the current behavior?

Template parts are completely undefined. Control authors and developers must search various places through code behind for these.

## What is the updated/expected behavior with this PR?

 * Added an attribute allowing each control to define their template parts. 
 * Added comments for the Pseduoclasses attribute while I was there
 * This does NOT define the template parts for controls -- that will have to be done in other PRs over time.

In the future such an attribute could be used by code generators to build strongly typed/named references to template parts in a uniform way. It is also possible for IDEs and designers to use this information in the future.

## How was the solution implemented (if it's not obvious)?

This is the same as what was done in WPF and the source code was copied and then modified from the WPF repo (MIT License).

https://github.com/dotnet/wpf/blob/a71d94d93b4af94061f6cd27c57b1d00efaccaab/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/TemplatePartAttribute.cs

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

None

## Obsoletions / Deprecations

None

## Fixed issues

Closes #7432

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019445-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Styling/Controls/Metadata/PseudoClassesAttribute.cs [hunk 1]
```diff
@@ -5,14 +5,27 @@
 
 namespace Avalonia.Controls.Metadata
 {
+    /// <summary>
+    /// Defines all pseudoclasses by name referenced and implemented by a control.
+    /// </summary>
+    /// <remarks>
+    /// This is currently used for code-completion in certain IDEs.
```
Comment (MarchingCube) [inline src/Avalonia.Styling/Controls/Metadata/PseudoClassesAttribute.cs:12]: FYI: Dev tools are also using this to generate checkboxes for toggling pseudo classes.

**Review by MarchingCube — APPROVED** [review]
Looks fine to me, this opens up IDE and dev tools integration.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019501-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by jmacato — APPROVED** [review]
code looks good. and a good way to add more metadata to help IDE's as well

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019503-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**